README
log4mongo-python
----------------
log4mongo-python is handler for MongoDB database for python logging, part of log4mongo.org project.
log4mongo-python is using pymongo driver - http://github.com/mongodb/mongo-python-driver

Requirements
------------

- python
- pymongo
- MongoDB

- tested against python 2.6.6+, MongoDB 1.6.4+, pymongo 1.9+

Configuration
-------------

Example handler python configuration::

    import logging
    from log4mongo.handlers import MongoHandler

    logger = logging.getLogger('test')
    logger.addHandler(MongoHandler(host='localhost'))
    logger.warning('test')

Author
------
char0n (Vladimir Gorej) <gorej@codescale.net>

References
----------
[1] http://www.mongodb.org/
[2] http://docs.python.org/library/logging.html
[3] http://github.com/mongodb/mongo-python-driver
[4] http://log4mongo.org

debian/changelog
python-log4mongo (1.0-1) UNRELEASED; urgency=medium

  * Initial package build

 -- Oz N Tiram <oz.tiram@gmail.com>  Wed, 24 May 2017 22:27:07 +0100


test.py
import unittest

from log4mongo.test import test_handlers

if __name__ == "__main__":
    loader = unittest.TestLoader()
    suite  = unittest.TestSuite()
    suite.addTests(loader.loadTestsFromModule(test_handlers))
    unittest.TextTestRunner(verbosity=2).run(suite)

tests/__init__.py
from test_mongo_handler import *

test/test_handlers.py
from log4mongo.handlers import (BufferedMongoHandler, MongoHandler,
                                write_method, write_many_method)
import log4mongo.handlers
from pymongo.errors import PyMongoError
import pymongo
if pymongo.version_tuple[0] >= 3:
    from pymongo.errors import ServerSelectionTimeoutError

try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO

import unittest
import logging
import time
import sys


class TestMongoHandler(unittest.TestCase):
    host_name = 'localhost'
    database_name = 'log4mongo_test'
    collection_name = 'logs_test'

    def setUp(self):
        self.handler = MongoHandler(host=self.host_name,
                                    database_name=self.database_name,
                                    collection=self.collection_name,
                                    )
        self.log = logging.getLogger('testLogger')
        self.log.setLevel(logging.DEBUG)
        self.log.addHandler(self.handler)
        self.old_stderr = sys.stdout
        sys.stderr = StringIO()

    def tearDown(self):
        self.handler.connection.drop_database(self.database_name)
        self.handler.close()
        self.log.removeHandler(self.handler)
        self.log = None
        self.handler = None
        sys.stderr.close()
        sys.stderr = self.old_stderr

    def test_connect(self):
        handler = MongoHandler(host='localhost',
                               database_name=self.database_name,
                               collection=self.collection_name)
        self.assertTrue(isinstance(handler, MongoHandler))
        self.handler.connection.drop_database(self.database_name)
        handler.close()

    def test_1_connect_failed(self):
        log4mongo.handlers._connection = None
        kwargs = {'connectTimeoutMS': 2000, 'serverselectiontimeoutms': 2000}
        if pymongo.version_tuple[0] < 3:
            with self.assertRaises(PyMongoError):
                MongoHandler(host='unknow_host',
                             database_name=self.database_name,
                             collection=self.collection_name,
                             **kwargs)
        else:
            with self.assertRaises(ServerSelectionTimeoutError):
                MongoHandler(host='unknow_host',
                             database_name=self.database_name,
                             collection=self.collection_name,
                             **kwargs)

    def test_connect_failed_silent(self):
        log4mongo.handlers._connection = None
        kwargs = {'connectTimeoutMS': 2000, 'serverselectiontimeoutms': 2000}
        handler = MongoHandler(host='unknow_host',
                               database_name=self.database_name,
                               collection=self.collection_name,
                               fail_silently=True,
                               **kwargs)
        self.assertTrue(isinstance(handler, MongoHandler))
        self.handler.connection.drop_database(self.database_name)
        handler.close()

    def test_emit(self):
        self.log.warning('test message')
        document = self.handler.collection.find_one(
            {'message': 'test message', 'level': 'WARNING'})
        self.assertEqual(document['message'], 'test message')
        self.assertEqual(document['level'], 'WARNING')

    def test_emit_exception(self):
        try:
            raise Exception('exc1')
        except:
            self.log.exception('test message')

        document = self.handler.collection.find_one(
            {'message': 'test message', 'level': 'ERROR'})
        self.assertEqual(document['message'], 'test message')
        self.assertEqual(document['level'], 'ERROR')
        self.assertEqual(document['exception']['message'], 'exc1')

    def test_emit_fail(self):
        self.handler.collection = ''
        self.log.warning('test warning')
        val = sys.stderr.getvalue()
        self.assertRegexpMatches(val, r"AttributeError: 'str' object has no attribute '{}'".format(write_method))

    def test_email_fail_silent(self):
        self.handler.fail_silently = True
        self.handler.collection = ''
        self.log.warning('test warming')
        self.assertEqual(sys.stderr.getvalue(), '')

    def test_contextual_info(self):
        self.log.info('test message with contextual info',
                      extra={'ip': '127.0.0.1', 'host': 'localhost'})
        document = self.handler.collection.find_one(
            {'message': 'test message with contextual info', 'level': 'INFO'})
        self.assertEqual(document['message'],
                         'test message with contextual info')
        self.assertEqual(document['level'], 'INFO')
        self.assertEqual(document['ip'], '127.0.0.1')
        self.assertEqual(document['host'], 'localhost')

    def test_contextual_info_adapter(self):
        adapter = logging.LoggerAdapter(self.log,
                                        {'ip': '127.0.0.1',
                                         'host': 'localhost'})
        adapter.info('test message with contextual info')
        document = self.handler.collection.find_one({'message': 'test message with contextual info', 'level': 'INFO'})
        self.assertEqual(document['message'], 'test message with contextual info')
        self.assertEqual(document['level'], 'INFO')
        self.assertEqual(document['ip'], '127.0.0.1')
        self.assertEqual(document['host'], 'localhost')


class TestCappedMongoHandler(TestMongoHandler):

    capped_max = 10

    def setUp(self):
        self.handler = MongoHandler(host=self.host_name, database_name=self.database_name,
                                    collection=self.collection_name, capped=True, capped_max=self.capped_max)
        self.log = logging.getLogger('testLogger')
        self.log.setLevel(logging.DEBUG)
        self.log.addHandler(self.handler)
        self.old_stderr = sys.stdout
        sys.stderr = StringIO()

    def test_capped(self):
        options = self.handler.db.command('collstats', self.collection_name)
        self.assertEqual(options['max'], 10)
        self.assertEqual(options['capped'], 1)

    def test_capped_max(self):
        for i in range(self.capped_max * 2):
            self.log.info('test capped info')
        documents = self.handler.collection.find()
        self.assertEqual(documents.count(), 10)

    def test_override_no_capped_collection(self):
        # Creating no capped handler
        self.handler_no_capped = MongoHandler(host=self.host_name,
                                              database_name=self.database_name,
                                              collection=self.collection_name)
        self.log.removeHandler(self.handler)
        self.log.addHandler(self.handler_no_capped)
        self.log.info('test info')
        # Creating capped handler
        self.handler_capped = MongoHandler(host=self.host_name,
                                           database_name=self.database_name,
                                           collection=self.collection_name,
                                           capped=True,
                                           capped_max=self.capped_max)
        self.log.addHandler(self.handler)
        self.log.info('test info')

    def test_override_capped_collection(self):
        # Creating capped handler
        self.handler_capped = MongoHandler(
            host=self.host_name, database_name=self.database_name,
            collection=self.collection_name, capped=True,
            capped_max=self.capped_max)
        self.log.removeHandler(self.handler)
        self.log.addHandler(self.handler)
        self.log.info('test info')
        # Creating no capped handler
        self.handler_no_capped = MongoHandler(host=self.host_name,
                                              database_name=self.database_name,
                                              collection=self.collection_name)
        self.log.addHandler(self.handler_no_capped)
        self.log.info('test info')

    def test_emit_on_reconnect_existing_capped_collection(self):
        self.log.removeHandler(self.handler)
        handler = MongoHandler(host=self.host_name, database_name=self.database_name,
                                    collection=self.collection_name, capped=True, capped_max=self.capped_max)
        self.log.addHandler(handler)

        self.log.warning('test message')
        document = handler.collection.find_one({'message': 'test message', 'level': 'WARNING'})
        self.assertEqual(document['message'], 'test message')
        self.assertEqual(document['level'], 'WARNING')


class TestBufferedMongoHandler(TestMongoHandler):
    collection_name = 'buffered_logs_test'

    def setUp(self):
        self.handler = BufferedMongoHandler(host=self.host_name,
                                    database_name=self.database_name,
                                    collection=self.collection_name,
                                    buffer_size=5, buffer_periodical_flush_timing=None,
                                    buffer_early_flush_level=logging.CRITICAL
                                    )
        self.log = logging.getLogger('testLogger')
        self.log.setLevel(logging.DEBUG)
        self.log.addHandler(self.handler)
        self.old_stderr = sys.stdout
        sys.stderr = StringIO()

    def force_flush(self):
        self.handler.flush_to_mongo()

    def test_contextual_info(self):
        self.log.info('test message with contextual info',
                      extra={'ip': '127.0.0.1', 'host': 'localhost'})
        self.force_flush()
        document = self.handler.collection.find_one(
            {'message': 'test message with contextual info', 'level': 'INFO'})
        self.assertEqual(document['message'],
                         'test message with contextual info')
        self.assertEqual(document['level'], 'INFO')
        self.assertEqual(document['ip'], '127.0.0.1')
        self.assertEqual(document['host'], 'localhost')

    def test_contextual_info_adapter(self):
        adapter = logging.LoggerAdapter(self.log,
                                        {'ip': '127.0.0.1',
                                         'host': 'localhost'})
        adapter.info('test message with contextual info')
        self.force_flush()
        document = self.handler.collection.find_one(
            {'message': 'test message with contextual info', 'level': 'INFO'})
        self.assertEqual(document['message'], 'test message with contextual info')
        self.assertEqual(document['level'], 'INFO')
        self.assertEqual(document['ip'], '127.0.0.1')
        self.assertEqual(document['host'], 'localhost')

    def test_emit(self):
        self.log.warning('test message')
        self.force_flush()
        document = self.handler.collection.find_one(
            {'message': 'test message', 'level': 'WARNING'})
        self.assertEqual(document['message'], 'test message')
        self.assertEqual(document['level'], 'WARNING')

    def test_emit_exception(self):
        try:
            raise Exception('exc1')
        except:
            self.log.exception('test message')
        self.force_flush()

        document = self.handler.collection.find_one(
            {'message': 'test message', 'level': 'ERROR'})
        self.assertEqual(document['message'], 'test message')
        self.assertEqual(document['level'], 'ERROR')
        self.assertEqual(document['exception']['message'], 'exc1')

    def test_emit_fail(self):
        self.handler.collection = ''
        self.log.warn('test warning')
        self.force_flush()
        val = sys.stderr.getvalue()
        self.assertRegexpMatches(val, r"AttributeError: 'str' object has no attribute '{}'".format(write_many_method))

    def test_buffer(self):
        self.force_flush()
        self.assertEqual(len(self.handler.buffer), 0, "Ensure buffer should be empty")

        self.log.warning('test_buffer message')

        document = self.handler.collection.find_one({'message': 'test_buffer message'})
        self.assertIsNone(document, "Should not have been written to database")
        self.assertEqual(len(self.handler.buffer), 1, "Buffer size should be 1")

        self.log.info('test_buffer message')
        self.log.debug('test_buffer message')
        self.log.info('test_buffer message')

        doc_amount = self.handler.collection.find({'message': 'test_buffer message'}).count()
        self.assertEqual(doc_amount, 0, "Nothing should have been written to database")
        self.assertEqual(len(self.handler.buffer), 4, "Buffer size should be 4")

        self.log.warning('test_buffer message')

        doc_amount = self.handler.collection.find({'message': 'test_buffer message'}).count()

        self.assertEqual(doc_amount, 5, "Buffer size reached, buffer should have been written to database")
        self.assertEqual(len(self.handler.buffer), 0, "Ensure buffer should be empty")

        self.log.warning('test_buffer message 2')

        document = self.handler.collection.find_one({'message': 'test_buffer message 2'})
        self.assertIsNone(document, 'Should not have been written to database')
        self.assertEqual(len(self.handler.buffer), 1, "Buffer size should be 1")

    def test_buffer_early_flush(self):
        self.force_flush()
        self.assertEqual(len(self.handler.buffer), 0, "Ensure buffer should be empty")

        self.log.info('test_buffer_early_flush message')

        document = self.handler.collection.find_one({'message': 'test_buffer_early_flush message'})
        self.assertIsNone(document, "Should not have been written to database")
        self.assertEqual(len(self.handler.buffer), 1, "Buffer size should be 1")

        self.log.critical('test_buffer_early_flush message')
        doc_amount = self.handler.collection.find({'message': 'test_buffer_early_flush message'}).count()
        self.assertEqual(doc_amount, 2, "2 messages should have been written to database")
        self.assertEqual(len(self.handler.buffer), 0, "Buffer should now be empty")

        doc_amount = self.handler.collection.find({'message': 'test_buffer_early_flush message',
                                                     'level': 'INFO'}).count()
        self.assertEqual(doc_amount, 1, "One INFO message should have been written to database")

        doc_amount = self.handler.collection.find({'message': 'test_buffer_early_flush message',
                                                     'level': 'CRITICAL'}).count()
        self.assertEqual(doc_amount, 1, "One CRITICAL message should have been written to database")

    def test_buffer_periodical_flush(self):

        # Creating capped handler
        self.handler_periodical = BufferedMongoHandler(host=self.host_name,
                                                       database_name=self.database_name,
                                                       collection=self.collection_name,
                                                       buffer_size=5, buffer_periodical_flush_timing=2.0,
                                                       buffer_early_flush_level=logging.CRITICAL)
        self.log.removeHandler(self.handler)
        self.log.addHandler(self.handler_periodical)

        self.log.info('test periodical buffer')
        document = self.handler_periodical.collection.find_one({'message': 'test periodical buffer'})
        self.assertIsNone(document, "Should not have been written to database")  # except if your computer is really slow
        self.assertEqual(len(self.handler_periodical.buffer), 1, "Buffer size should be 1")

        self.log.info('test periodical buffer')
        document = self.handler_periodical.collection.find_one({'message': 'test periodical buffer'})
        self.assertIsNone(document, "Should not have been written to database")
        self.assertEqual(len(self.handler_periodical.buffer), 2, "Buffer size should be 2")

        time.sleep(2.5)  # wait a bit so the periodical timer thread has kicked off the buffer flush

        document = self.handler_periodical.collection.find_one({'message': 'test periodical buffer'})
        self.assertIsNotNone(document, "Should not have been written to database")
        doc_amount = self.handler_periodical.collection.find({'message': 'test periodical buffer'}).count()
        self.assertEqual(doc_amount, 2, "Should have found 2 documents written to database")
        self.assertEqual(len(self.handler_periodical.buffer), 0, "Buffer should be empty")

        self.assertTrue(self.handler_periodical.buffer_timer_thread.is_alive())
        self.handler_periodical.destroy()
        time.sleep(0.2)  # waiting a tiny bit so that the child thread actually exits
        self.assertFalse(self.handler_periodical.buffer_timer_thread.is_alive(),
                         "Child timer thread should be dead by now. Slow computer?")

        # reset to previous
        self.log.removeHandler(self.handler_periodical)
        self.log.addHandler(self.handler)


logging4mongo.py
import logging
from bson.timestamp import Timestamp
import traceback
import cStringIO
from pymongo import Connection

class MongoHandler(logging.Handler):

    def __init__(self, host='localhost', port=27017, database_name='logging4mongo', collection='logs', username=None, password=None):
        logging.Handler.__init__(self)
        self.host           = host
        self.port           = port
        self.database_name  = database_name
        self.collectionName = collection
        self.username       = username
        self.password       = password

        self.connection = None
        self.collection = None

        self._connect()

    def format(self, record):
        document = self._logRecordToDict(record)

        if record.exc_info is not None:
            document.update({
                'exception' : self._exceptionToDict(record.exc_info)
            })

        return document

    def _connect(self):
        try:
            self.connection = Connection(host=self.host, port=self.port)
            db = self.connection[self.database_name]
            if self.username is not None and self.password is not None:
                db.authenticate(self.username, self.password)
            self.collection = db[self.collectionName]
        except:
            print 'Silent handler, server connecting or collection selecting failed'

    def close(self):
        if self.connection is not None:
            self.connection.disconnect()

    def _logRecordToDict(self, record):
        return {
            'timestamp'  : Timestamp(int(record.created), int(record.msecs)),
            'level'      : record.levelname,
            'thread'     : record.thread,
            'message'    : record.getMessage(),
            'fileName'   : record.pathname,
            'method'     : record.funcName,
            'lineNumber' : record.lineno,
            'className'  : record.__class__.__name__
        }

    def _exceptionToDict(self, e):
        sio = cStringIO.StringIO()
        traceback.print_exception(e[0], e[1], e[2], None, sio)
        s = sio.getvalue()
        sio.close()
        if s[-1:] == "\n":
            s = s[:-1]

        return {
            'message'    : str(e[1]),
            'code'       : 0,
            'stactTrace' : s
        }




    def emit(self, record):
        if self.collection is not None:
            try:
                self.collection.save(self.format(record))
            except:
                self.handleError(record)


if __name__ == '__main__':
    log = logging.getLogger('')
    log.addHandler(logging.StreamHandler())
    log.addHandler(MongoHandler())


    def test():
        try:
            raise Exception('test', 'test1')
        except Exception:
            log.exception('exception catched')
            pass

    test()




provision_vagrant.sh
#!/usr/bin/env bash
export DEBIAN_FRONTEND=noninteractive

sudo locale-gen "en_US.UTF-8"
sudo dpkg-reconfigure locales

echo "LC_ALL=en_US.UTF-8" >> /etc/environment
echo "LANG=en_US.UTF-8" >> /etc/environment

PACKAGES="python-dev python3-dev build-essential"


apt-get update
apt-get install -y ${PACKAGES}

if [ ! -f  /usr/local/bin/pip ]; then
    wget https://bootstrap.pypa.io/get-pip.py
	sudo python get-pip.py
fi

if [ ! -f  /usr/local/bin/pip3 ]; then
	sudo python3 get-pip.py
fi


PYTHON_PACKAGES="pymongo>=2.8 pexpect coverage unittest2 \
virtualenvwrapper"

sudo pip3 install ${PYTHON_PACKAGES}
sudo pip install ${PYTHON_PACKAGES}

# install mongodb

sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 614D985504A2163B
echo "deb http://repo.mongodb.org/apt/ubuntu \
	"$(lsb_release -sc)"/mongodb-org/3.1 multiverse" \
	| sudo tee /etc/apt/sources.list.d/mongodb-org-3.1.list

sudo apt-get update
sudo apt-get install -y mongodb-org-unstable





log4mongo/handlers.py
import datetime as dt
import logging

try:
    from pymongo import MongoClient as Connection
except ImportError:
    from pymongo import Connection

from pymongo.collection import Collection
from pymongo.errors import OperationFailure, PyMongoError
import pymongo
if pymongo.version_tuple[0] >= 3:
    from pymongo.errors import ServerSelectionTimeoutError
    write_method = 'insert_one'
    write_many_method = 'insert_many'
else:
    write_method = 'save'
    write_many_method = 'insert'

"""
Example format of generated bson document:
{
    'thread': -1216977216,
    'threadName': 'MainThread',
    'level': 'ERROR',
    'timestamp': datetime.datetime(2016, 8, 16, 15, 20, 24, 794341),
    'message': 'test message',
    'module': 'test_module',
    'fileName': '/var/projects/python/log4mongo-python/tests/test_handlers.py',
    'lineNumber': 38,
    'method': 'test_emit_exception',
    'loggerName':  'testLogger',
    'exception': {
        'stackTrace': 'Traceback (most recent call last):
                       File "/var/projects/python/log4mongo-python/tests/\
                           test_handlers.py", line 36, in test_emit_exception
                       raise Exception(\'exc1\')
                       Exception: exc1',
        'message': 'exc1',
        'code': 0
    }
}
"""
_connection = None


class MongoFormatter(logging.Formatter):

    DEFAULT_PROPERTIES = logging.LogRecord(
        '', '', '', '', '', '', '', '').__dict__.keys()

    def format(self, record):
        """Formats LogRecord into python dictionary."""
        # Standard document
        document = {
            'timestamp': dt.datetime.utcnow(),
            'level': record.levelname,
            'thread': record.thread,
            'threadName': record.threadName,
            'message': record.getMessage(),
            'loggerName': record.name,
            'fileName': record.pathname,
            'module': record.module,
            'method': record.funcName,
            'lineNumber': record.lineno
        }
        # Standard document decorated with exception info
        if record.exc_info is not None:
            document.update({
                'exception': {
                    'message': str(record.exc_info[1]),
                    'code': 0,
                    'stackTrace': self.formatException(record.exc_info)
                }
            })
        # Standard document decorated with extra contextual information
        if len(self.DEFAULT_PROPERTIES) != len(record.__dict__):
            contextual_extra = set(record.__dict__).difference(
                set(self.DEFAULT_PROPERTIES))
            if contextual_extra:
                for key in contextual_extra:
                    document[key] = record.__dict__[key]
        return document


class MongoHandler(logging.Handler):

    def __init__(self, level=logging.NOTSET, host='localhost', port=27017,
                 database_name='logs', collection='logs',
                 username=None, password=None, authentication_db='admin',
                 fail_silently=False, formatter=None, capped=False,
                 capped_max=1000, capped_size=1000000, reuse=True, **kwargs):
        """
        Setting up mongo handler, initializing mongo database connection via
        pymongo.

        If reuse is set to false every handler will have it's own MongoClient.
        This could hammer down your MongoDB instance, but you can still use
        this option.

        The default is True. As such a program with multiple handlers
        that log to mongodb will have those handlers share a single connection
        to MongoDB.
        """
        logging.Handler.__init__(self, level)
        self.host = host
        self.port = port
        self.database_name = database_name
        self.collection_name = collection
        self.username = username
        self.password = password
        self.authentication_database_name = authentication_db
        self.fail_silently = fail_silently
        self.connection = None
        self.db = None
        self.collection = None
        self.authenticated = False
        self.formatter = formatter or MongoFormatter()
        self.capped = capped
        self.capped_max = capped_max
        self.capped_size = capped_size
        self.reuse = reuse
        self._connect(**kwargs)

    def _connect(self, **kwargs):
        """Connecting to mongo database."""
        global _connection
        if self.reuse and _connection:
            self.connection = _connection
        else:
            if pymongo.version_tuple[0] < 3:
                try:
                    self.connection = Connection(host=self.host,
                                                 port=self.port, **kwargs)
                # pymongo >= 3.0 does not raise this error
                except PyMongoError:
                    if self.fail_silently:
                        return
                    else:
                        raise
            else:
                self.connection = Connection(host=self.host, port=self.port,
                                             **kwargs)
                try:
                    self.connection.is_primary
                except ServerSelectionTimeoutError:
                    if self.fail_silently:
                        return
                    else:
                        raise
            _connection = self.connection

        self.db = self.connection[self.database_name]
        if self.username is not None and self.password is not None:
            auth_db = self.connection[self.authentication_database_name]
            self.authenticated = auth_db.authenticate(self.username,
                                                      self.password)

        if self.capped:
            #
            # We don't want to override the capped collection
            # (and it throws an error anyway)
            try:
                self.collection = Collection(self.db, self.collection_name,
                                             capped=True, max=self.capped_max,
                                             size=self.capped_size)
            except OperationFailure:
                # Capped collection exists, so get it.
                self.collection = self.db[self.collection_name]
        else:
            self.collection = self.db[self.collection_name]

    def close(self):
        """
        If authenticated, logging out and closing mongo database connection.
        """
        if self.authenticated:
            self.db.logout()
        if self.connection is not None:
            self.connection.close()

    def emit(self, record):
        """Inserting new logging record to mongo database."""
        if self.collection is not None:
            try:
                getattr(self.collection, write_method)(self.format(record))
            except Exception:
                if not self.fail_silently:
                    self.handleError(record)

    def __exit__(self, type, value, traceback):
        self.close()


class BufferedMongoHandler(MongoHandler):

    def __init__(self, level=logging.NOTSET, host='localhost', port=27017,
                 database_name='logs', collection='logs',
                 username=None, password=None, authentication_db='admin',
                 fail_silently=False, formatter=None, capped=False,
                 capped_max=1000, capped_size=1000000, reuse=True,
                 buffer_size=100, buffer_periodical_flush_timing=5.0,
                 buffer_early_flush_level=logging.CRITICAL, **kwargs):
        """
        Setting up buffered mongo handler, initializing mongo database connection via
        pymongo.

        This subclass aims to provide buffering mechanism to avoid hammering the database server and
        write-locking the database too often (even if mongo is performant in that matter).

        If buffer_periodical_flush_timer is set to None or 0, no periodical flush of the buffer will be done.
        It means that buffered messages might be stuck here for a while until the buffer full or
        a critical message is sent (both causing flush).

        If buffer_periodical_flush_timer is set to numeric value, a thread with timer will be launched
        to call the buffer flush periodically.
        """

        MongoHandler.__init__(self, level=level, host=host, port=port, database_name=database_name, collection=collection,
                              username=username, password=password, authentication_db=authentication_db,
                              fail_silently=fail_silently, formatter=formatter, capped=capped, capped_max=capped_max,
                              capped_size=capped_size, reuse=reuse, **kwargs)
        self.buffer = []
        self.buffer_size = buffer_size
        self.buffer_periodical_flush_timing = buffer_periodical_flush_timing
        self.buffer_early_flush_level = buffer_early_flush_level
        self.last_record = None #kept for handling the error on flush
        self.buffer_timer_thread = None

        self._buffer_lock = None
        self._timer_stopper = None

        # setup periodical flush
        if self.buffer_periodical_flush_timing:

            # clean exit event
            import atexit
            atexit.register(self.destroy)

            # retrieving main thread as a safety
            import threading
            main_thead = threading.current_thread()
            self._buffer_lock = threading.RLock()

            # call at interval function
            def call_repeatedly(interval, func, *args):
                stopped = threading.Event()

                # actual thread function
                def loop():
                    while not stopped.wait(interval) and main_thead.is_alive():  # the first call is in `interval` secs
                        func(*args)

                timer_thread = threading.Thread(target=loop)
                timer_thread.daemon = True
                timer_thread.start()
                return stopped.set, timer_thread

            # launch thread
            self._timer_stopper, self.buffer_timer_thread = call_repeatedly(self.buffer_periodical_flush_timing, self.flush_to_mongo)

    def emit(self, record):
        """Inserting new logging record to buffer and flush if necessary."""

        self.add_to_buffer(record)

        if len(self.buffer) >= self.buffer_size or record.levelno >= self.buffer_early_flush_level:
            self.flush_to_mongo()
        return

    def buffer_lock_acquire(self):
        """Acquire lock on buffer (only if periodical flush is set)."""
        if self._buffer_lock:
            self._buffer_lock.acquire()

    def buffer_lock_release(self):
        """Release lock on buffer (only if periodical flush is set)."""
        if self._buffer_lock:
            self._buffer_lock.release()

    def add_to_buffer(self, record):
        """Add a formatted record to buffer."""

        self.buffer_lock_acquire()

        self.last_record = record
        self.buffer.append(self.format(record))

        self.buffer_lock_release()

    def flush_to_mongo(self):
        """Flush all records to mongo database."""
        if self.collection is not None and len(self.buffer) > 0:
            self.buffer_lock_acquire()
            try:

                getattr(self.collection, write_many_method)(self.buffer)
                self.empty_buffer()

            except Exception as e:
                if not self.fail_silently:
                    self.handleError(self.last_record) #handling the error on flush
            finally:
                self.buffer_lock_release()

    def empty_buffer(self):
        """Empty the buffer list."""
        del self.buffer
        self.buffer = []

    def destroy(self):
        """Clean quit logging. Flush buffer. Stop the periodical thread if needed."""
        if self._timer_stopper:
            self._timer_stopper()
        self.flush_to_mongo()
        self.close()


